### ISSUE TITLE
[2.5.9] rename with destination hostname!part looses part

### ISSUE THREAD

**Issue by MASHtm** [cyrusimap/cyrus-imapd#46]
doing a 

```
rename user.bla user.bla destination.tld!partxx
```

looses the partxx part within cmd_rename and ends up on the same destination partition every time.

the code splits "location" into "destserver" and "destpart" by overwriting the "!" with \0.
and afterwards it uses 'location' with XFER and RENAME. 'destpart' is missing completely from XFER.

**Comment by MASHtm** [comment]
since all moves ended up on part08 while part01 is my default-partition... default-partition is ignored as well. part08 was not the partition with most available space... at least at the moment it run full.

All my backends have
```
defaultpartition: part01

partition-part01: /var/spool/imap/part01
...
```
for part01-part18

**Comment by MASHtm** [comment]
this codesection is completely broken up to master IMO. Best part is
```
            if (c) {
                *c++ = '\0';
                destpart = xstrdupnull(c);
            } else {
                destpart = xstrdup(location);
                free(destserver);
            }

            if (destserver) {
                if (destpart && !strcmp(destserver,config_servername)) {
                    // XFER
                    prot_printf(s->out,
                            "%s XFER \"%s\" \"%s\" %s\r\n",
                            tag,
                            oldname,
                            newname,
                            location
                        );
```

free(destserver)? => if (destserver) => strcmp(destserver)? 
config_servername instead of mbentry->server?
XFER Syntax broken?

a first proposal with all comments left intact
```
--- imapd.c.orig	2016-10-05 12:58:19.607783414 +0200
+++ imapd.c	2016-10-05 13:44:32.171449243 +0200
@@ -6079,16 +6079,22 @@
 	    char *destpart = NULL;
 
 	    destserver = xstrdupnull(location);
-	    c = strchr(location, '!');
+	    // do not modify location, used later
+	    //c = strchr(location, '!');
+	    c = strchr(destserver, '!');
 	    if (c) {
+		// destserver ends before "!" => hostname
 		*c++ = '\0';
+		// destpart after "!"
 		destpart = xstrdupnull(c);
-	    } else {
-		destpart = xstrdup(location);
-		free(destserver);
-	    }
+	    } //else {
+		// no partition => use destserver as is
+		// destpart = xstrdup(location);
+		// free? and used in the next codeline? 
+		// free(destserver);
+	    //}
 
-	    if (destserver) {
+/*	    if (destserver) {
 		if (destpart && !strcmp(destserver,config_servername)) {
 		    // XFER
 		    prot_printf(s->out,
@@ -6109,6 +6115,38 @@
 			    location
 			);
 		}
+	    } // (destserver) */
+
+	    if (destserver && !strcmp(destserver, mbentry->server)) {
+	    // same host? => RENAME
+	    // what happens if no part is given?
+		prot_printf(s->out,
+		        "%s RENAME \"%s\" \"%s\" %s\r\n",
+		        tag,
+		        oldname,
+		        newname,
+		        location
+		    );
+	    } else {
+	    // different host? => XFER 
+		if (destpart) {
+		    // XFER
+		    prot_printf(s->out,
+			    "%s XFER \"%s\" %s %s\r\n",
+			    tag,
+			    oldname,
+			    destserver,
+			    destpart
+			);
+		} else {
+		    // XFER
+		    prot_printf(s->out,
+			    "%s XFER \"%s\" %s\r\n",
+			    tag,
+			    oldname,
+			    destserver
+			);
+		}
 	    } // (destserver)
 
 	    res = pipe_until_tag(s, tag, 0);
```

**Comment by MASHtm** [comment]
at least the part for XFER with part works for me

**Comment by MASHtm** [comment]
IMO parts of this bugreport are true for 3.0 as well. essentially all but the "free before use" for destserver case.

master code uses config_servername as well instead of the correct mbentry->server which is wrong for murder frontends.

rename user.x user.x newhost
is handled wrong as well since it results in a
RENAME
on the old backend what is wrong as I already reported for 2.5

So please fix this essential murder functionality at least before releasing 3.0 since you didn't even comment on my bug report for 2.5

**Comment by elliefm** [comment]
Thanks for bumping this. I think I was working on unit tests for murder at the time and set this aside, then got sidetracked, and in the meantime the issue got lost under the imported bugzilla stuff... sorry!

Grabbing the assignment so I don't forget it again, will look at this properly on friday :)

**Comment by elliefm** [comment]
What's the expected behaviour for a case like `rename user.x user.y newhost!newpart`?

Your patch looks like it forgets the "user.y" aspect, and only moves to the new host/partition, but keeping the existing name

Do we expect that when rename has a location specifed, that the old and new mailbox names should be identical (and only the location changes)?

A quick scan of RFC 3501 suggests the location argument is an extension -- I'm not sure if it's cyrus-specific or a standard one though?

**Comment by elliefm** [comment]
Oh, I see this later on, cool: `"%s NO Cross-server or cross-partition move w/rename not supported\r\n",`

**Comment by elliefm** [comment]
Work in progress here: https://github.com/cyrusimap/cyrus-imapd/compare/cyrus-imapd-2.5...elliefm:v25/46-murder-rename

Are you able to try it out?

**Comment by MASHtm** [comment]
I was at least able to take a look at your patch. Looks good for all cases. The case "same server, different partition" is now proxied to the current backend with "rename". 2.4 denied that at all. But it should work since the backend will deny anyway if the proxy user doesn't have the rights to do it.

Until April I'll not have time to test it in our environment.

**Comment by elliefm** [comment]
No worries, thanks for looking at it.  I'll keep trying to get a viable murder setup into our test suite

**Comment by elliefm** [comment]
The problem I'm bumping into at the moment seems to be that once the XFER has been proxied from the frontend to the original backend, it's authenticated as the proxy user rather than an admin user.  Which means the XFER then fails, due to insufficient permissions to create the mailbox on the new backend.

**Comment by MASHtm** [comment]
that's correct. the proxy user used by the backends needs to be admin user on all backends.

proxy_authname, proxy_password and admins need to be set accordingly on all backends.

**Comment by elliefm** [comment]
Great, that helps a lot, thanks :)

2.5 branch tidied up (same link as before), and equivalent patches against master are at: https://github.com/cyrusimap/cyrus-imapd/compare/master...elliefm:v30/46-murder-rename

Cassandane test here: https://github.com/cyrusimap/cassandane/compare/master...elliefm:murder-multi-backends

The Cassandane test works okay in isolation, but it's failing during full suite runs: when it selects the mailbox on the old backend to verify that it's gone, the select unexpectedly succeeds.  Not sure what the underlying cause is yet.

**Comment by elliefm** [comment]
Delving further (on master):

* When run in isolation, the test succeeds, correctly.

* When run as part of the full suite, the rename operation fails but returns an 'ok' status (below).  The test continues on to verifying that the move happens, and discovers that it didn't.

```
C: 4 rename user.cassandane user.cassandane localhost:9128!
S: * NO USER cassandane (The remote Server(s) denied the operation)
S: 4 OK Completed
```

So a couple of things:

* The test needs to also look for untagged NO responses, not just the tagged response
* Why on earth is the rename behaving differently here

**Comment by elliefm** [comment]
Conversely, on 2.5:

* When run in isolation, the test succeeds, correctly.

* When run as part of the full suite, the rename operation fails and returns a 'no' status (below).

```
C: 4 rename user.cassandane user.cassandane localhost:9101!
S: 4 NO The remote Server(s) denied the operation
```

So, it's behaving in the same way, just the problem is noticed/reported at a slightly different point.  Again: what on earth

**Comment by elliefm** [comment]
... turns out an unrelated test was leaking its config overrides back into the defaults, so the test for this case would fail if they were run in the same invocation.

Fixes are now on the master and 2.5 branches, and cassandane tests too.  We now also have cassandane infrastructure for murder tests across multiple backends, so that will help nail down this sort of thing better in the future.

**Comment by elliefm** [comment]
This will be in 2.5.11

**Comment by xinrag** [comment]
This bug is still not fixed. In a murder of cyrus-imapd-3.0.3 and on a 2.5.11 backend I renamed a mailbox. During this process the mailbox got mooved to the default partition.
How to reproduce:
1. setup a simple murder...
2. configure more than one partition in a backend server.
3. use cyradm to create a mailbox on a non default partition (f.e. here we have "00" as default and 0[1-9] as further partitions... "cm user.test3 05"
4. to verify: "info user.test3" shoes the right partition 05
5. now rename... "renm user.test3 user.test33"
6. verify and see the failure: "info user.test33" ... "partition: 00"

Our 2.4.x backend servers hasn't this failure/problem.

Btw - "funny" behavior of the rename command : If I use the same name, no rename... just use another partition "renm user.test33 user.test33 09" works, everything is fine. If I try to to change the name and stay on the partition "renm user.test33 user.test36 09" I got the message "renamemailbox: Cross-server or cross-partition move w/rename not supported". Maybe the command isn't possible atm, ok. But at least the simple rename should work on the same partition.

**Comment by ksmurchison** [comment]
Just to be clear,

rename user.bla user.bla destination.tld!partxx

worked (before I left) and still works at CMU, who is running some flavor of 3.x

I'd argue that doing a rename across servers without specifying a destination partition SHOULD use the default partition.  There is no guarantee that the partition layout on different servers is the same.

A rename within a single server, SHOULD remain on the same partition unless specified otherwise.

I think that covers all of the use cases.

**Comment by ksmurchison** [comment]
I will look into renaming a user on the same backend

**Comment by elliefm** [comment]
I think the newer part of this (rename on same server gets moved to default partition) is the same thing that was fixed as #2907 for 3.0 and master just recently.

At this point I'm not gonna bother backporting to 2.5, but if someone else wants to do it and submit a PR I'd be happy to shepherd it through.
